README.md
<p align="center">
    <a href="https://github.com/unclezs/NovelHarverster/actions/workflows/gradle.yml">
    <img src="https://img.shields.io/github/workflow/status/unclezs/uncle-novel/Java%20CI%20with%20Gradle" alt="gradle build"/>
    </a>
    <a href="https://travis-ci.com/unclezs/uncle-novel">
    <img src="https://img.shields.io/travis/com/unclezs/uncle-novel/main?logo=travis" alt="Travis Build"/>
    </a>
    <a href="https://github.com/unclezs/jfx-launcher/blob/main/LICENSE">
    <img src="https://img.shields.io/github/license/unclezs/uncle-novel?color=%2340C0D0&label=License" alt="GitHub license"/>
    </a>
	<img src="https://img.shields.io/github/downloads/unclezs/uncle-novel/total" alt=""/>
	<img src="https://img.shields.io/badge/openjdk-11-green" alt=""/>
	<img src="https://img.shields.io/badge/platform-win linux mac-green" alt=""/>
</p>

# Uncle 小说

一个运行在 PC 端的应用，提供了全网小说的转码阅读功能。其核心为目录解析功能，辅以书源模式进行（全网小说）小说的内容获取。

本项目仅供学习交流，请勿用于商业用途，软件内产生的数据请关闭软件后立即删除！！

## 主要功能

- 搜书文本小说
- 搜书有声小说
- 全网搜书
- 文本小说书架
- 文本小说阅读器
- 有声小说书架
- 解析下载
- 下载管理
- 书源管理
- 软件设置
- 全局热键
- 主题定制
- 国际化支持
- 备份与恢复（WebDav）

## 预览

### 书架

书架提供了书籍的分组管理功能，同时支持按照分组批量更新书籍，也可以导入本地 TXT 小说到书架。

<img width="600" src="https://github.com/unclezs/uncle-novel/raw/main/app/packager/screenshot/home.png"/>

### 阅读器

<img width="600" src="https://github.com/unclezs/uncle-novel/raw/main/app/packager/screenshot/read.png"/>
<img width="600" src="https://github.com/unclezs/uncle-novel/raw/main/app/packager/screenshot/read1.png"/>

### 软件设置

<img width="600" src="https://github.com/unclezs/uncle-novel/raw/main/app/packager/screenshot/setting.png"/>


## 开发

### 本地点火

准备环境:

- `jdk` 至少需要 JDK11 版本。
- `npm install -g sass` 安装 sass，本项目使用 saas 将 scss 编译为 css。
- `sqlite3`，非必须，如果想重建 sqlite 数据库则需要安装。

然后直接运行 :app:runApp 任务即可。

```shell
./graldew :app:runApp
```

### 打包

```shell
# window 64 位包
./gradlew :app:packageWin64

# window 32 位包
./gradlew :app:packageWin32

# mac 包，如果开发机不是 aarch64(m1) 的则有可能不能在 aarch64 的 mac 上运行
./gradlew :app:packageMac

# linux 包，未测试..应该是不完善的
./gradlew :app:packageLinux
```

## 支持开发者~

<img src="https://cdn.unclezs.com/20210105090216.jpeg" alt=""/>



changelog
V5.0.39 2021-7-16
1.适配 Mac M1

V5.0.38 2021-7-16
1.修复阅读器字体与主页字体不一致

V5.0.37 2021-7-14
1.修复Mac闪退问题
2.修复解析下载在多页时点击下载卡死问题

V5.0.36 2021-7-13
1.修复下载一直在等待中问题
2.修复导出规则时正则转义错误

V5.0.35 2021-7-11
1.新增设置可直接使用系统代理
2.修复预处理脚本错误提示未生效问题

V5.0.34 2021-7-11
1.解析下载右键增加全选与全不选
2.小说书架书籍右键增加解析下载

V5.0.33 2021-7-9
1.修复JsonPath匹配异常

V5.0.32 2021-7-8
1.新增章节排序支持自定义脚本

V5.0.31 2021-7-7
1.修复TTS朗读上一页问题
2.增强书源调试器
3.书源规则在列表时支持规则类型混用

V5.0.30 2021-7-4
1.新增右键清空下载历史
2.正文规则新增下载延迟
3.搜索规则详情页采用GET请求

V5.0.29 2021-7-1
1.修复搜索规则必须同时指定详情页与搜索页问题

V5.0.28 2021-6-30
1.修复规则序列化异常问题
2.有声小说播放器优化

V5.0.27 2021-6-30
1.阅读器新增繁体简体互转化
2.阅读器增加是否启用动画开关

V5.0.26 2021-6-30
1.导出时简化书源
2.全网搜书增加返回上页按钮

V5.0.25 2021-6-29
1.优化动态网页获取

V5.0.24 2021-6-29
1.修复解析正文失败问题

V5.0.23 2021-6-29
1.完善规则编辑器
2.增加执行定义请求预处理脚本

V5.0.22 2021-6-29
1.修复退出后自动删除下载目录

V5.0.21 2021-6-29
1.优化解析正文翻页
2.修复无法自动获取小说标题问题
3.修复段落模式异常

V5.0.20 2021-6-28
1.修复有声书架目录右键菜单无效

V5.0.19 2021-6-28
1.优化规则解析器中正文请求次数
2.修复JsonPath规则解析后会带双引号
3.修复搜索与解析时使用书源不一致问题

V5.0.18 2021-6-28
1.修复书架书籍更新后章节顺序与解析时不一致
2.书源管理增加网络导入与剪贴板导入

V5.0.17 2021-6-27
1.修复任务下载失败后无法被删除
2.修复书架中书籍更新后目录乱序

V5.0.16 2021-6-26
1.修复替换净化规则正则无效

V5.0.15 2021-6-26
1.修复详情规则无法匹配问题

V5.0.14 2021-6-26
1.修复Mobi格式生成失败问题
2.修复书源编辑器的一些问题

V5.0.13 2021-6-26
1.调整默认为不进行分卷下载
2.调整替换规则默认不为正则，启用正则需要加regex:前缀
3.脚本工具支持匹配列表

V5.0.12 2021-6-25
1.升级kindlegen版本为V2.9
2.修复书源管理书源过多显示异常问题
3.修复书架书籍丢失问题

V5.0.11 2021-6-25
1.修复书架书籍封面只能更换一次问题
2.新增书架书籍封面右键手动获取
3.新增正文繁体转简体功能
4.修复最小化后再按老板键窗口无法显示

V5.0.10 2021-6-25
1.修复书源编辑器中权重无法编辑问题
2.修复预处理脚本部分异常问题

V5.09 2021-6-24
1.增加ESC退出阅读器
2.修复Epub章节名称重复
3.增加预处理脚本的初始params变量

V5.08 2021-6-24
1.修复无法生成Mobi格式文件
2.增加预处理脚本调试工具
3.修复阅读器窗口大小无法保存问题

V5.07 2021-6-24
1.修复搜索规则参数编译问题
2.修复章节规则调试器的信息不准确问题

V5.06 2021-6-24
1.修复字体无法显示问题
2.增加自定义软件字体
3.增加书架书籍标题总是显示设置项

V5.05 2021-6-23
1.修复默认翻页规则未生效
2.修复搜索规则参数编译失败问题

V5.04 2021-6-23
1.修复规则编辑器参数编译异常
2.乱码问题可以通过更改系统默认字体解决

V5.03 2021-6-23
1.修复阅读器快捷键不能使用问题
2.添加托盘菜单
3.增加阅读器滚轮翻页

V5.02 2021-6-22
1.修复规则预处理脚本异常问题
2.尝试修复部分设备乱码问题
3.修复字体加载失败导致程序崩溃问题

V5.01 2021-6-16
1.V5.0测试版本发布


src/main/resources/conf/xmSign.js
// tslint:disable

//const De = cb => {
//  return new Promise(cb);
//};

//function De{
//    return new Promise(cb);
//}
//
//function de{
//    return new Promise(cb);
//}
//const De = function(cb){
//    return new Promise(cb);
//}

//
//function de(){
//    return new Promise(cb);
//}
function Ve() {
  var t = 0 < arguments.length && void 0 !== arguments[0] && arguments[0];
  Ue() ||
    (window.XM_SERVER_CLOCK && !t) ||
    De(function(t, e) {
      var n = new XMLHttpRequest();
      n.open("GET", "https://www.ximalaya.com/revision/time", !0),
        n.send(null),
        (n.onreadystatechange = function() {
          if (4 === n.readyState) {
            var e = Number(n.responseText);
            (e = isNaN(e) ? Date.now() : e), t(e);
          }
        });
    }).then(function(t) {
      (We = t), (window.XM_SERVER_CLOCK = t), $e();
    });



}
//
//function Ve() {
//  var t = 0 < arguments.length && void 0 !== arguments[0] && arguments[0];
//  Ue() ||(window.XM_SERVER_CLOCK && !t);
//    var n = new XMLHttpRequest();
//      n.open("GET", "https://www.ximalaya.com/revision/time", !0),
//        n.send(null),
//        (n.onreadystatechange = function() {
//          if (4 === n.readyState) {
//            var e = Number(n.responseText);
//            (e = isNaN(e) ? Date.now() : e), (We = t), (window.XM_SERVER_CLOCK = t);
//          }
//        });
//
//
//}

function Be(t) {
  return (
    !!t.constructor &&
    "function" == typeof t.constructor.isBuffer &&
    t.constructor.isBuffer(t)
  );
}
function Ue() {
  return "undefined" == typeof window;
}
function Fe(t) {
  return ~~(Math.random() * t);
}

var e, n;
(e = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"),
  (n = {
    rotl: function(t, e) {
      return (t << e) | (t >>> (32 - e));
    },
    rotr: function(t, e) {
      return (t << (32 - e)) | (t >>> e);
    },
    endian: function(t) {
      if (t.constructor == Number)
        return (16711935 & n.rotl(t, 8)) | (4278255360 & n.rotl(t, 24));
      for (var e = 0; e < t.length; e++) t[e] = n.endian(t[e]);
      return t;
    },
    randomBytes: function(t) {
      for (var e = []; 0 < t; t--) e.push(Math.floor(256 * Math.random()));
      return e;
    },
    bytesToWords: function(t) {
      for (var e = [], n = 0, r = 0; n < t.length; n++, r += 8)
        e[r >>> 5] |= t[n] << (24 - (r % 32));
      return e;
    },
    wordsToBytes: function(t) {
      for (var e = [], n = 0; n < 32 * t.length; n += 8)
        e.push((t[n >>> 5] >>> (24 - (n % 32))) & 255);
      return e;
    },
    bytesToHex: function(t) {
      for (var e = [], n = 0; n < t.length; n++)
        e.push((t[n] >>> 4).toString(16)), e.push((15 & t[n]).toString(16));
      return e.join("");
    },
    hexToBytes: function(t) {
      for (var e = [], n = 0; n < t.length; n += 2)
        e.push(parseInt(t.substr(n, 2), 16));
      return e;
    },
    bytesToBase64: function(t) {
      for (var n = [], r = 0; r < t.length; r += 3)
        for (
          var o = (t[r] << 16) | (t[r + 1] << 8) | t[r + 2], i = 0;
          i < 4;
          i++
        )
          8 * r + 6 * i <= 8 * t.length
            ? n.push(e.charAt((o >>> (6 * (3 - i))) & 63))
            : n.push("=");
      return n.join("");
    },
    base64ToBytes: function(t) {
      t = t.replace(/[^A-Z0-9+\/]/gi, "");
      for (var n = [], r = 0, o = 0; r < t.length; o = ++r % 4)
        0 != o &&
          n.push(
            ((e.indexOf(t.charAt(r - 1)) & (Math.pow(2, -2 * o + 8) - 1)) <<
              (2 * o)) |
              (e.indexOf(t.charAt(r)) >>> (6 - 2 * o))
          );
      return n;
    }
  });

var Ne = n;

var Le = {
    utf8: {
      stringToBytes: function(t) {
        return Le.bin.stringToBytes(unescape(encodeURIComponent(t)));
      },
      bytesToString: function(t) {
        return decodeURIComponent(escape(Le.bin.bytesToString(t)));
      }
    },
    bin: {
      stringToBytes: function(t) {
        for (var e = [], n = 0; n < t.length; n++)
          e.push(255 & t.charCodeAt(n));
        return e;
      },
      bytesToString: function(t) {
        for (var e = [], n = 0; n < t.length; n++)
          e.push(String.fromCharCode(t[n]));
        return e.join("");
      }
    }
  },
  Ie = Le;
function je(t) {
  return (
    null != t &&
    (Be(t) ||
      (function(t) {
        return (
          "function" == typeof t.readFloatLE &&
          "function" == typeof t.slice &&
          Be(t.slice(0, 0))
        );
      })(t) ||
      !!t._isBuffer)
  );
}
//
var He = (function(t) {
    var t = {
      exports: {}
    };
    var e, n, r, o, i;
    (e = Ne),
      (n = Ie.utf8),
      (r = je),
      (o = Ie.bin),
      ((i = function(t, a) {
        t.constructor == String
          ? (t =
              a && "binary" === a.encoding
                ? o.stringToBytes(t)
                : n.stringToBytes(t))
          : r(t)
          ? (t = Array.prototype.slice.call(t, 0))
          : Array.isArray(t) || (t = t.toString());
        for (
          var u = e.bytesToWords(t),
            s = 8 * t.length,
            c = 1732584193,
            l = -271733879,
            f = -1732584194,
            p = 271733878,
            h = 0;
          h < u.length;
          h++
        )
          u[h] =
            (16711935 & ((u[h] << 8) | (u[h] >>> 24))) |
            (4278255360 & ((u[h] << 24) | (u[h] >>> 8)));
        (u[s >>> 5] |= 128 << s % 32), (u[14 + (((64 + s) >>> 9) << 4)] = s);
        var d = i._ff,
          v = i._gg,
          y = i._hh,
          g = i._ii;
        for (h = 0; h < u.length; h += 16) {
          var m = c,
            b = l,
            w = f,
            E = p;
          (l = g(
            (l = g(
              (l = g(
                (l = g(
                  (l = y(
                    (l = y(
                      (l = y(
                        (l = y(
                          (l = v(
                            (l = v(
                              (l = v(
                                (l = v(
                                  (l = d(
                                    (l = d(
                                      (l = d(
                                        (l = d(
                                          l,
                                          (f = d(
                                            f,
                                            (p = d(
                                              p,
                                              (c = d(
                                                c,
                                                l,
                                                f,
                                                p,
                                                u[h + 0],
                                                7,
                                                -680876936
                                              )),
                                              l,
                                              f,
                                              u[h + 1],
                                              12,
                                              -389564586
                                            )),
                                            c,
                                            l,
                                            u[h + 2],
                                            17,
                                            606105819
                                          )),
                                          p,
                                          c,
                                          u[h + 3],
                                          22,
                                          -1044525330
                                        )),
                                        (f = d(
                                          f,
                                          (p = d(
                                            p,
                                            (c = d(
                                              c,
                                              l,
                                              f,
                                              p,
                                              u[h + 4],
                                              7,
                                              -176418897
                                            )),
                                            l,
                                            f,
                                            u[h + 5],
                                            12,
                                            1200080426
                                          )),
                                          c,
                                          l,
                                          u[h + 6],
                                          17,
                                          -1473231341
                                        )),
                                        p,
                                        c,
                                        u[h + 7],
                                        22,
                                        -45705983
                                      )),
                                      (f = d(
                                        f,
                                        (p = d(
                                          p,
                                          (c = d(
                                            c,
                                            l,
                                            f,
                                            p,
                                            u[h + 8],
                                            7,
                                            1770035416
                                          )),
                                          l,
                                          f,
                                          u[h + 9],
                                          12,
                                          -1958414417
                                        )),
                                        c,
                                        l,
                                        u[h + 10],
                                        17,
                                        -42063
                                      )),
                                      p,
                                      c,
                                      u[h + 11],
                                      22,
                                      -1990404162
                                    )),
                                    (f = d(
                                      f,
                                      (p = d(
                                        p,
                                        (c = d(
                                          c,
                                          l,
                                          f,
                                          p,
                                          u[h + 12],
                                          7,
                                          1804603682
                                        )),
                                        l,
                                        f,
                                        u[h + 13],
                                        12,
                                        -40341101
                                      )),
                                      c,
                                      l,
                                      u[h + 14],
                                      17,
                                      -1502002290
                                    )),
                                    p,
                                    c,
                                    u[h + 15],
                                    22,
                                    1236535329
                                  )),
                                  (f = v(
                                    f,
                                    (p = v(
                                      p,
                                      (c = v(
                                        c,
                                        l,
                                        f,
                                        p,
                                        u[h + 1],
                                        5,
                                        -165796510
                                      )),
                                      l,
                                      f,
                                      u[h + 6],
                                      9,
                                      -1069501632
                                    )),
                                    c,
                                    l,
                                    u[h + 11],
                                    14,
                                    643717713
                                  )),
                                  p,
                                  c,
                                  u[h + 0],
                                  20,
                                  -373897302
                                )),
                                (f = v(
                                  f,
                                  (p = v(
                                    p,
                                    (c = v(
                                      c,
                                      l,
                                      f,
                                      p,
                                      u[h + 5],
                                      5,
                                      -701558691
                                    )),
                                    l,
                                    f,
                                    u[h + 10],
                                    9,
                                    38016083
                                  )),
                                  c,
                                  l,
                                  u[h + 15],
                                  14,
                                  -660478335
                                )),
                                p,
                                c,
                                u[h + 4],
                                20,
                                -405537848
                              )),
                              (f = v(
                                f,
                                (p = v(
                                  p,
                                  (c = v(c, l, f, p, u[h + 9], 5, 568446438)),
                                  l,
                                  f,
                                  u[h + 14],
                                  9,
                                  -1019803690
                                )),
                                c,
                                l,
                                u[h + 3],
                                14,
                                -187363961
                              )),
                              p,
                              c,
                              u[h + 8],
                              20,
                              1163531501
                            )),
                            (f = v(
                              f,
                              (p = v(
                                p,
                                (c = v(c, l, f, p, u[h + 13], 5, -1444681467)),
                                l,
                                f,
                                u[h + 2],
                                9,
                                -51403784
                              )),
                              c,
                              l,
                              u[h + 7],
                              14,
                              1735328473
                            )),
                            p,
                            c,
                            u[h + 12],
                            20,
                            -1926607734
                          )),
                          (f = y(
                            f,
                            (p = y(
                              p,
                              (c = y(c, l, f, p, u[h + 5], 4, -378558)),
                              l,
                              f,
                              u[h + 8],
                              11,
                              -2022574463
                            )),
                            c,
                            l,
                            u[h + 11],
                            16,
                            1839030562
                          )),
                          p,
                          c,
                          u[h + 14],
                          23,
                          -35309556
                        )),
                        (f = y(
                          f,
                          (p = y(
                            p,
                            (c = y(c, l, f, p, u[h + 1], 4, -1530992060)),
                            l,
                            f,
                            u[h + 4],
                            11,
                            1272893353
                          )),
                          c,
                          l,
                          u[h + 7],
                          16,
                          -155497632
                        )),
                        p,
                        c,
                        u[h + 10],
                        23,
                        -1094730640
                      )),
                      (f = y(
                        f,
                        (p = y(
                          p,
                          (c = y(c, l, f, p, u[h + 13], 4, 681279174)),
                          l,
                          f,
                          u[h + 0],
                          11,
                          -358537222
                        )),
                        c,
                        l,
                        u[h + 3],
                        16,
                        -722521979
                      )),
                      p,
                      c,
                      u[h + 6],
                      23,
                      76029189
                    )),
                    (f = y(
                      f,
                      (p = y(
                        p,
                        (c = y(c, l, f, p, u[h + 9], 4, -640364487)),
                        l,
                        f,
                        u[h + 12],
                        11,
                        -421815835
                      )),
                      c,
                      l,
                      u[h + 15],
                      16,
                      530742520
                    )),
                    p,
                    c,
                    u[h + 2],
                    23,
                    -995338651
                  )),
                  (f = g(
                    f,
                    (p = g(
                      p,
                      (c = g(c, l, f, p, u[h + 0], 6, -198630844)),
                      l,
                      f,
                      u[h + 7],
                      10,
                      1126891415
                    )),
                    c,
                    l,
                    u[h + 14],
                    15,
                    -1416354905
                  )),
                  p,
                  c,
                  u[h + 5],
                  21,
                  -57434055
                )),
                (f = g(
                  f,
                  (p = g(
                    p,
                    (c = g(c, l, f, p, u[h + 12], 6, 1700485571)),
                    l,
                    f,
                    u[h + 3],
                    10,
                    -1894986606
                  )),
                  c,
                  l,
                  u[h + 10],
                  15,
                  -1051523
                )),
                p,
                c,
                u[h + 1],
                21,
                -2054922799
              )),
              (f = g(
                f,
                (p = g(
                  p,
                  (c = g(c, l, f, p, u[h + 8], 6, 1873313359)),
                  l,
                  f,
                  u[h + 15],
                  10,
                  -30611744
                )),
                c,
                l,
                u[h + 6],
                15,
                -1560198380
              )),
              p,
              c,
              u[h + 13],
              21,
              1309151649
            )),
            (f = g(
              f,
              (p = g(
                p,
                (c = g(c, l, f, p, u[h + 4], 6, -145523070)),
                l,
                f,
                u[h + 11],
                10,
                -1120210379
              )),
              c,
              l,
              u[h + 2],
              15,
              718787259
            )),
            p,
            c,
            u[h + 9],
            21,
            -343485551
          )),
            (c = (c + m) >>> 0),
            (l = (l + b) >>> 0),
            (f = (f + w) >>> 0),
            (p = (p + E) >>> 0);
        }
        return e.endian([c, l, f, p]);
      })._ff = function(t, e, n, r, o, i, a) {
        var u = t + ((e & n) | (~e & r)) + (o >>> 0) + a;
        return ((u << i) | (u >>> (32 - i))) + e;
      }),
      (i._gg = function(t, e, n, r, o, i, a) {
        var u = t + ((e & r) | (n & ~r)) + (o >>> 0) + a;
        return ((u << i) | (u >>> (32 - i))) + e;
      }),
      (i._hh = function(t, e, n, r, o, i, a) {
        var u = t + (e ^ n ^ r) + (o >>> 0) + a;
        return ((u << i) | (u >>> (32 - i))) + e;
      }),
      (i._ii = function(t, e, n, r, o, i, a) {
        var u = t + (n ^ (e | ~r)) + (o >>> 0) + a;
        return ((u << i) | (u >>> (32 - i))) + e;
      }),
      (i._blocksize = 16),
      (i._digestsize = 16);
    t.exports = function(t, n) {
      if (null == t) throw new Error("Illegal argument " + t);
      var r = e.wordsToBytes(i(t, n));
      return n && n.asBytes
        ? r
        : n && n.asString
        ? o.bytesToString(r)
        : e.bytesToHex(r);
    };
    return t.exports;
  })(),
  We = null,
  ze = null,
  $e = function() {
    ze = setInterval(function() {
      window.XM_SERVER_CLOCK - We <= 18e4
        ? (window.XM_SERVER_CLOCK += 3333)
        : (clearInterval(ze), Ve(!(ze = null)));
    }, 3333);
  };

//Ve();

function java(d){
    return (
    He("ximalaya-" + d) + "(" + Fe(100) + ")" + d + "(" + Fe(100) + ")" + getnow()
    );
}



function getnow(){
    var timestamp=new Date().getTime();
    return timestamp;
}
